[BD-14] Add filtering and search support to library APIs

mavidser opened the pull request:
This PR adds filtering support to some libraries API (libraries_v2_list and libraries_v2_blocks_list APIs namely).
- For `GET /libraries/v2/`: Add an org filter, and a text_search parameter which performs partial matching on Title, ID, Org, and Description.
- Index XBlocks in elasticsearch
- For `GET /libraries/v2/{lib_key_str}/blocks/`: Add a text_search parameter which performs partial matching on the ID and display name.

**JIRA Ticket**: [BLENDED-526](https://openedx.atlassian.net/browse/BLENDED-526), [Opencraft/SE-2930](https://tasks.opencraft.com/browse/SE-2930), [Opencraft/SE-2938](https://tasks.opencraft.com/browse/SE-2938)

**Testing instructions**:

To test APIs:
- From `make studio-shell` run: `./manage.py cms reindex_content_library --all` to index all existing libraries.
- Make calls to the `GET /api/libraries/v2/` and `GET /libraries/v2/{lib_key_str}/blocks/` APIs and use different values for supported filters. You can use `/swagger` on studio to make the requests.

Run tests:
- From the blockstore directory, run `make testserver` to start a test-specific instance.
- From `make studio-shell` run: `EDXAPP_ENABLE_ELASTICSEARCH_FOR_TESTS=1  EDXAPP_RUN_BLOCKSTORE_TESTS=1 python -Wd -m pytest --ds=cms.envs.test cms/djangoapps/contentstore/tests/test_libraries.py openedx/core/djangoapps/content_libraries/tests/test_content_libraries.py openedx/core/djangoapps/content_libraries/tests/test_libraries_index.py`

**Reviewers:**
- [ ] @arbrandes
- [ ] edX reviewer[s] TBD

**Author notes:**
- This PR also includes commits from #24530, as it builds up on indexes introduced in there.

edx-webhook commented:
Thanks for the pull request, @mavidser! I've created [OSPR-4827](https://openedx.atlassian.net/browse/OSPR-4827) to keep track of it in JIRA, where we prioritize reviews.  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Please let us know once your PR is ready for our review and all tests are green.


<!-- jenkins ok to test -->

natabene commented:
@mavidser Thank you for your contribution. Please let me know once it is ready for our review.

kdmccormick commented:
Thanks @mavidser . I'll add reviewing this to my queue.

arbrandes commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -410,35 +503,69 @@ def delete_library(library_key):
+        except (ConnectionError) as e:
+            log.exception(e)
+
+    if metadata is None:
```
What if an indexed search returns no hits?  It looks like we'll be returning all items (and from the slow API), which is probably not the desired behavior.  What is the downside to using an "else", here?  Are you protecting for unindexed libraries?

mavidser replied:
If the indexed search returns no hits, `metadata` would be an empty list `[]`, and skip this block.
Using this check instead of `else` so that this block gets executed if indexing is enabled, but connection to elasticsearch fails (cf. the no-op on `ConnectionError` in the line above)

arbrandes replied:
Ah, of course!  Makes total sense. +1

arbrandes commented on openedx/core/djangoapps/content_libraries/libraries_index.py:
```diff
@@ -0,0 +1,302 @@
+
+class LibraryBlockIndexer(SearchIndexerBase):
+    """
+    Class to perform indexing for blockstore-based content libraries
```
Nit: maybe the doc should be at least a little different from the ContentLibraryIndexer's one.

arbrandes commented on openedx/core/djangoapps/content_libraries/views.py:
```diff
@@ -61,20 +63,78 @@ def wrapped_fn(*args, **kwargs):
+    )
     def get(self, request):
         """
         Return a list of all content libraries that the user has permission to
         view. This is a temporary view for development and returns at most 50
         libraries.
```
This docstring is no longer accurate. :)

arbrandes commented:
@mavidser, I've tested this and it works fine.  Code LGTM, too.  +1!

mavidser commented:
@kdmccormick This branch should now be ready for a first review now. I've also introduced elasticsearch into the tests as efficient searching and filtering needed some complex queries which the mock engine wouldn't have been able to test. This is enabled by a flag (mentioned in the testing instructions of the description) right now as we don't have a separate target to run elastic tests yet. Reusing the devstack elastic for testing at the time.

Also, I'm still considering django migrations as a way to update elastic schemas, else version upgrades could become destructive to the APIs (empty index=no results). Any thoughts on this? cc @bradenmacdonald

bradenmacdonald commented:
@mavidser Using django migrations for elastic schema/data updates sounds good to me, yeah. Any other mechanism requires doing something manual to update them, and it won't get done consistently.

Please review https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations though - all the same concerns will apply to ElasticSearch migrations as apply to django migrations. In particular, renaming a column or deleting one must be done a very specific way, as two versions of code will be accessing the index at the same time.

bradenmacdonald commented:
Also thanks for adding an ADR 👏

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/docs/decisions/0001-index-libraries-in-elasticsearch.rst:
```diff
@@ -0,0 +1,57 @@
+-------
+
+The new content libraries reside in blockstore instead of edx-platform's models,
+which means that we are no longer able to query databases to get metadata
```
Maybe say "to get **complete** metadata" because _some_ metadata is in the LMS database, and some in the Blockstore MySQL database.

kdmccormick commented:
@mavidser Could you rebase? The state of this branch seems incompatible with the requirements installed into the latest edxapp docker image.

mavidser commented:
@kdmccormick rebased. :+1: 

@bradenmacdonald - thanks for the docs - will keep blue-green deployments in mind, Might have to change how IDs and schema versions are stored to make that work though. I think it's best to do that in a separate PR just because this one has gotten big enough for review.

kdmccormick reviewed (commented):
Thanks for the ADR! The decisions in it sound good to me.

> I've also introduced elasticsearch into the tests as efficient searching and filtering needed some complex queries which the mock engine wouldn't have been able to test. This is enabled by a flag (mentioned in the testing instructions of the description) right now as we don't have a separate target to run elastic tests yet. Reusing the devstack elastic for testing at the time.

This runs against the grain of existing edx-platform unit tests, which AFAIK do not require external services. Furthermore, I don't think our Jenkins CI setup would support running the tests that require Elasticsearch without modification. I encourage you to take another swing at testing your code without relying on real ES. Or, help me understand why we might need ES present to thoroughly test this code. 

> I think it's best to do that in a separate PR just because this one has gotten big enough for review.

Looking into Django migrations seems worthwhile, but I agree with merging this first :+1:

mavidser commented:
@kdmccormick Ack. Yeah, to be fair elastic is useful only when developing and testing [this part of the code](https://github.com/open-craft/edx-platform/blob/0223396a0e0c078ce2d3e0544d5dad1b3c75c330/openedx/core/djangoapps/content_libraries/libraries_index.py#L145-L177). It can probably be mocked away, but I found elastic quite eager to error out with slight changes in queries.

It is enabled by an explicit flag, but now that it works, I don't mind mocking it and assuming safe execution (one can always find ways to test features when working on that query). [Here's the specific changes done for allowing elastic in tests](https://github.com/edx/edx-platform/pull/24510/commits/0c20753e0d17a1a87d6b432ec857d9e49034527d), which can be easily reverted or minimised further.

bradenmacdonald commented:
@kdmccormick @mavidser  Should we put something in https://github.com/edx/edx-e2e-tests or https://github.com/edx/cypress-e2e-tests that tests the end-to-end search functionality using real ElasticSearch then?

kdmccormick commented:
@bradenmacdonald @mavidser I would avoid adding these tests to `edx-e2e-tests`. Those tests end up having a much higher flakiness rate and maintenance burden than our unit tests, so we've been trying reserve them for highly critical interactions like basic courseware usage and certificate purchases. If Elasticsearch were to become involved in the rendering of XBlocks in LMS, then that may be worth writing an e2e test for; however, if library listing page in Studio broke, it would be fairly harmless to just let that come in as a bug report.

It'd be ideal if it were easy enough to write and maintain end-to-end tests for the platform such that it was standard practice to add them along with new features, but we're just not there yet.

mavidser commented:
@kdmccormick Sure, that sounds fair. Should I then keep the tests as it is, or remove the elasticsearch-enabling flags from the testenv? I'm inclined to at least keep the [`elasticsearch_test` decorator](https://github.com/edx/edx-platform/pull/24510/commits/0c20753e0d17a1a87d6b432ec857d9e49034527d#diff-9f784f60f30c9f8682b4c2e1100f921dR46-R57) to make it easier to test by manually overriding the env.

kdmccormick commented:
@mavidser The decorator looks good to keep. You could even move the `ELASTIC_SEARCH_CONFIG` override into the decorator if you wanted to. Please just leave a comment that this isn't used unless explicitly enabled locally.

Mainly, I want to keep the logic out of `cms/envs/test.py`, partially because it won't be used by default and thus makes the file noisier, and partially because adopting the newly-merged [OEP-45](https://open-edx-proposals.readthedocs.io/en/latest/oep-0045-arch-ops-and-config.html#configuration) will require us to remove all conditional logic from our settings files anyway.

Let me know when you're ready for another pass from me!

mavidser commented:
@kdmccormick Thanks for the OEP link - I've removed the logic from the `cms/envs/test.py` and moved it to the decorator instead. This should be ready for your review post @arbrandes's code-review :)

arbrandes commented on openedx/core/djangoapps/content_libraries/libraries_index.py:
```diff
@@ -25,116 +28,234 @@
+        # NOTE: Increment ContentLibraryIndexer.SCHEMA_VERSION if the following schema is updated to avoid dealing
+        # with outdated indexes which might cause errors due to missing/invalid attributes.
+        return {
+            "schema_version": ContentLibraryIndexer.SCHEMA_VERSION,
```
Great idea!  +1

arbrandes reviewed (approved):
Just gave this one last pass.  Looks great, @mavidser.  I have no changes to request.

@kdmccormick, as far as I'm concerned, this is good to merge.

kdmccormick reviewed (approved):
Looks good! Great work!

Given that only LabXChange relies on V2 APIs currently, I will leave it up to you folks to assess the merge risk.

I am happy to merge this as early as tomorrow (Friday) morning.

mavidser commented:
@kdmccormick Sounds good. Note that indexing stays disabled in this PR too, and can be enabled once we figure out dealing with migrations in elastic.

kdmccormick commented:
@mavidser Just to be sure, is that a "Sounds good, go ahead and merge" or "Sounds good, we'll let you know when we want to merge"? :)

mavidser commented:
Ah, it's good to merge this PR. :+1:

kdmccormick commented:
Great! Should be in production by noon EDT. If you folks at OpenCraft notice this is causing any problems, I'll be watching #edx-opencraft until ~5pm EDT.

kdmccormick commented:
@mavidser This hasn't gone out yet. Our edx-platform deployment pipeline was stalled today, and we've purposely paused it for the rest of the weekend. This should go out Monday morning.

kdmccormick commented:
@mavidser @arbrandes Assuming we don't run into any more pipeline hiccups, this should be in production around 9:30am EDT.

kdmccormick commented:
Just kidding--an end-to-end test failure is currently blocking deployment. Let's just say "between 10am and noon EDT".

kdmccormick commented:
Between 1:30-2:30pm EDT. It hasn't been a good couple of days for CD.

arbrandes commented:
> **EdX Release Notice**: This PR has been rolled back from the production environment.

Has there been an issue I'm (we're?) not aware of, @kdmccormick?

bradenmacdonald commented:
@arbrandes 90% of the time, those messages are unrelated, due to someone else's PR or a deployment/pipeline issue.

kdmccormick commented:
Right. The nature of our pipeline is that changes will sometimes go out and/or need to be rolled back in batches. This change just happened to be batched with an unrelated bugged change, which as you can see, we've been thrashing over a bit. I believe we're in the clear, now, though.

arbrandes commented:
Ah, of course - I should've realized this.  Been seeing the release tags on git for years. ;)
